tck/src/main/java/org/reactivestreams/tck/SubscriberBlackboxVerification.java

```diff
@@ -96,11 +96,25 @@ public void required_spec201_blackbox_mustSignalDemandViaSubscriptionRequest() t
       @Override
       public void run(BlackboxTestStage stage) throws InterruptedException {
         triggerRequest(stage.subProxy().sub());
-        final long n = stage.expectRequest();// assuming subscriber wants to consume elements...
+        final long requested = stage.expectRequest();// assuming subscriber wants to consume elements...
+        final long signalsToEmit = Math.min(requested, 512); // protecting against Subscriber which sends ridiculous large demand
 
         // should cope with up to requested number of elements
-        for (int i = 0; i < n; i++)
+        for (int i = 0; i < signalsToEmit && sampleIsCancelled(stage, i, 10); i++)
           stage.signalNext();
+        
+        // we complete after `signalsToEmit` (which can be less than `requested`), 
+        // which is legal under https://github.com/reactive-streams/reactive-streams-jvm#1.2
+        stage.sendCompletion();
```

Comment:
I guess the "polite" implementation would check if it is not cancelled before issuing the completion. (so that the implementation does not rely on onComplete if cancelled)